src/dataflow/components/dataflow-program.tsx

```diff
@@ -245,15 +259,23 @@ export class DataflowProgram extends BaseComponent<IProps, IState> {
 
       this.intervalHandle = setInterval(this.heartBeat, HEARTBEAT_INTERVAL);
 
+      const isProgramRunning = this.isProgramRunning();
+      this.setState({isProgramRunning});
     })();
   }
 
+  private isProgramRunning = () => {
+    return (Boolean(this.props.programId) && this.props.programEndTime > Date.now());
+  }
+
   private runProgram = () => {
     const programData: any = this.generateProgramData();
     uploadProgram(programData);
     this.setState({isProgramRunning: true});
   }
   private stopProgram = () => {
+    const programEndTime = Date.now();
+    this.props.onSetProgramEndTime(programEndTime);
```

Comment:
Let's not forget that we'll want to send a message to the server setting the new end time, if the user stops early. But it doesn't need to be a part of this PR.

Reply:
Do we have something in place now that we can call?